### PR TITLE
Feat/4 application crud

### LINKED ISSUE #4: Application CRUD endpoints on backend
Endpoints:

GET /applications
- This should return all the applications and it should be pageable.
- The list returned needs to be sorted by `createdAt`
- For pagination it should accept a `page` parameter for the current page index (starting from 0) and a `page_size` parameter for the number of items on each page
- The backend needs to check if the current user has permission to call the endpoint

GET /applications/my
- Returns the application that belongs to the user in the current application period
- It somewhat collides with the next endpoint, but all the applicationId-s are numbers, so it shouldn't be a problem

GET /applications/{applicationId}
- The backend needs to check if the current user has permission to call the endpoint

POST /applications
- Creates a new application
- The backend needs to check if there is already an application (maybe a DB constraint would be the best way)

PATCH  /applications/{applicationId}
- Modifies an existing application
- The backend needs to check if the current user has permission to do it

### PR DESCRIPTION
**Pull request by FearsomeRover**
Closes #4

### CODE REVIEW

File: apps/backend/src/application/application.service.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+              authSchId: user.authSchId,
+            },
+          },
+          applicationPeriod: {
+            connect: {
+              id: createApplicationDto.applicationPeriodId,
+            },
+          },
+        },
+      });
```
Comment (Tschonti) [inline apps/backend/src/application/application.service.ts:35]: I think we should check whether the applicationPeriod is still accepting new applications

File: apps/backend/src/application/application.controller.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  @Get('my')
+  getCurrentUserApplication(@CurrentUser() user: User): Promise<Application | null> {
+    return this.applicationService.getCurrentUserApplication(user);
+  }
+
+  @UseGuards(AuthGuard('jwt'), RolesGuard)
+  @ApiBearerAuth()
+  @Roles(Role.BODY_ADMIN, Role.BODY_MEMBER)
+  @Get(':id')
+  findOne(@Param('id') id: string): Promise<Application> {
```
Comment (Tschonti) [inline apps/backend/src/application/application.controller.ts:44]: ParseIntPipe()

File: apps/backend/src/application/application.controller.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    return this.applicationService.findOne(Number(id));
+  }
+
+  @UseGuards(AuthGuard('jwt'), RolesGuard)
+  @ApiBearerAuth()
+  @Roles(Role.BODY_ADMIN, Role.BODY_MEMBER)
+  @Patch(':id')
+  update(@Param('id') id: string, @Body() updateApplicationDto: UpdateApplicationDto): Promise<Application> {
```
Comment (Tschonti) [inline apps/backend/src/application/application.controller.ts:52]: ParseIntPipe()

File: apps/backend/src/application/application.controller.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    return this.applicationService.update(Number(id), updateApplicationDto);
+  }
+
+  @UseGuards(AuthGuard('jwt'), RolesGuard)
+  @ApiBearerAuth()
+  @Delete(':id')
+  remove(@Param('id') id: string, @CurrentUser() user: User): Promise<Application> {
```
Comment (Tschonti) [inline apps/backend/src/application/application.controller.ts:59]: ParseIntPipe()

File: apps/backend/src/application/application.service.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+      if (e instanceof BadRequestException) {
+        throw e;
+      }
+      throw new BadRequestException('Nem sikerült létrehozni');
+    }
+  }
+
+  @UseGuards(AuthGuard('jwt'), RolesGuard)
+  @ApiBearerAuth()
+  @Roles(Role.BODY_ADMIN, Role.BODY_MEMBER)
```
Comment (Tschonti) [inline apps/backend/src/application/application.service.ts:61]: These are useless here

File: apps/backend/src/application/application.service.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  async findAll() {
+    return await this.prisma.application.findMany();
+  }
```
Comment (Tschonti) [inline apps/backend/src/application/application.service.ts:64]: Body members are probably only interested in the applications for the current (or last) period. applicationPeriodId could be an optional parameter if they want to look at older data

File: apps/backend/src/application/application.service.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+  async findOne(id: number): Promise<Application> {
+    const application = this.prisma.application.findFirst({
+      where: {
+        id,
+      },
+    });
```
Comment (Tschonti) [inline apps/backend/src/application/application.service.ts:71]: not sure about this, but it's possible that it's more efficient to use findUnique if you're querying by unique fields

File: apps/backend/src/application/application.service.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    }
+    if (application.userId === user.authSchId || user.role === Role.BODY_ADMIN || user.role === Role.BODY_MEMBER) {
+      return await this.prisma.application.delete({
+        where: {
+          id,
+        },
+      });
+    }
+    throw new ForbiddenException('Nem törölheted mások jelentkezését');
+  }
```
Comment (Tschonti) [inline apps/backend/src/application/application.service.ts:138]: should the user be able to delete their application after the application period has ended? that could cause some confusion for the body members

File: apps/backend/src/application/dto/create-application.dto.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+export class CreateApplicationDto extends OmitType(Application, [
+  'id',
+  'status',
+  'user',
+  'userId',
+  'applicationPeriod',
+  'createdAt',
+  'updatedAt',
+]) {}
```
Comment (Tschonti) [inline apps/backend/src/application/dto/create-application.dto.ts:13]: it's probably better to just list that one property in this case, took me a minute to find which one is it xd. And some validation would be nice

**Review by Tschonti — CHANGES_REQUESTED** [review]

File: apps/backend/src/application/application.controller.ts (hunk 4 above)
Comment (Tschonti) [reply apps/backend/src/application/application.controller.ts:65]: ez kimaradt xd

File: apps/backend/src/application/dto/update-application.dto.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+import { ApplicationStatus } from '@prisma/client';
+
+export class UpdateApplicationDto {
+  applicationStatus: ApplicationStatus;
+}
```
Comment (Tschonti) [inline apps/backend/src/application/dto/update-application.dto.ts:5]: either validate the enum here (might have to create a new enum for that, not sure if class-validator can handle the prisma enum), or handle the db error if invalid enum was sent

Comment (FearsomeRover) [reply apps/backend/src/application/dto/update-application.dto.ts:5]: It was fine with the prisma enums

**Review by Tschonti — APPROVED** [review]

**Pull request merged**